File: dec8/one.mjs
import fs from "fs/promises";

(async () => {
    const file = (await fs.readFile("./input.txt", "utf-8")).trim();
    let lines = file.split("\n");
    const map = new Map();
    lines.forEach((line, i) => {
        line.split("").forEach((char, j) => {
            if(char !== "."){
                map.get(char) ? map.get(char).push([j, i]) : map.set(char, [[j, i]]);
            }
        });
    });
    const set = new Set();
    const max_x = lines[0].length;
    const max_y = lines.length;
    for(let char of map){
        const entry = char[1];
        for(let i = 0; i < entry.length; i++){
            for(let j = 0; j < entry.length; j++){
                if(i !== j){
                    let x, y;
                    x = entry[i][0] + (entry[i][0] - entry[j][0]);
                    y = entry[i][1] + (entry[i][1] - entry[j][1]);
                    if(x >= 0 && x < max_x && y >= 0 && y < max_y) set.add(`${x},${y}`);
                }
            }
        }
    }
    console.log(set.size);
})()

File: dec8/one.rs
use std::fs;
use std::collections::{HashMap, HashSet};

fn main() {
    let file = fs::read_to_string("input.txt").expect("Failed to read file").trim().to_string();
    let lines: Vec<&str> = file.lines().collect();
    let mut map: HashMap<char, Vec<(isize, isize)>> = HashMap::new();

    for (i, line) in lines.iter().enumerate() {
        for (j, ch) in line.chars().enumerate() {
            if ch != '.' {
                map.entry(ch).or_insert(Vec::new()).push((j as isize, i as isize));
            }
        }
    }

    let mut set: HashSet<(isize, isize)> = HashSet::new();
    let max_x = lines[0].len() as isize;
    let max_y = lines.len() as isize;

    for positions in map.values() {
        for i in 0..positions.len() {
            for j in 0..positions.len() {
                if i != j {
                    let (x1, y1) = positions[i];
                    let (x2, y2) = positions[j];

                    let x = x1 + (x1 - x2);
                    let y = y1 + (y1 - y2);

                    if x >= 0 && x < max_x && y >= 0 && y < max_y {
                        set.insert((x, y));
                    }
                }
            }
        }
    }

    println!("{}", set.len());
}